src/vm/jithelpers.cpp

```diff
@@ -2464,7 +2464,7 @@ BOOL ObjIsInstanceOf(Object *pObject, TypeHandle toTypeHnd)
         CALL_MANAGED_METHOD(fCast, BOOL, args);
         INDEBUG(managedType = NULL); // managedType isn't protected during the call
 
-        if (exception != NULL)
+        if (!fCast && canThrowCustomInvalidCast && exception != NULL)
```

Comment:
BTW: This design does not allow efficient implementation of isinst because of one has to always allocate exception on failure. It would be better to pass the bool flag as argument to the ICastable method, and let it take care of throwing the exception. We should clean in up before ICastable becomes a real public contract (not necessarily as part of this change).